File: backend/layer2bridge/full_node_interface.py
from typing import List, Dict, Any, Optional
from bitcoin_core_rpc import (
    BitcoinRPCClient,
    LoadWalletResponse,
    ListSinceBlockResponse,
    GetTransactionResponse,
    AddressGroupingItem,
    GetBlockHeaderResponse
)
from config_models.models import Layer2BridgeBitcoinConfFileSettings
from OnboardingLogger import OnboardingLogger

DEFAULT_TESTNET = True
TESTNET_TARGETCONFIRMATIONS = 3
MAINNET_TARGETCONFIRMATIONS = 6
SATOSHI_PER_BITCOIN = 100000000

class BitcoinRPC:
    def __init__(self, rpc_config: Layer2BridgeBitcoinConfFileSettings, wallet_name: str, testnet: bool = DEFAULT_TESTNET):
        self.client = BitcoinRPCClient(rpc_config, wallet_name)
        self.testnet = testnet
        self.wallet_name = wallet_name
        OnboardingLogger(f"rpc_port: {rpc_config.rpcport}")

    def getMinimumTransactionAmount(self) -> int:
        return 1000

    def getTargetConfirmations(self) -> int:
        return TESTNET_TARGETCONFIRMATIONS if self.testnet else MAINNET_TARGETCONFIRMATIONS

    async def loadWallet(self) -> LoadWalletResponse:
        try:
            response = await self.client.loadwallet(self.wallet_name)
            if response.result:
                return response.result
            if response.is_wallet_already_loaded:
                return LoadWalletResponse(name=self.wallet_name, warning="Already loaded")
            if response.error:
                raise Exception(f"Failed to load wallet: {response.error.message}")
            raise Exception("Failed to load wallet: Unknown error")
        except Exception as e:
            OnboardingLogger(e)
            raise

    async def broadcastTransaction(self, pendingWithdrawals: Dict[str, Any]) -> str:
        amounts = {}
        subtractfeefrom = []
        for pendingWithdrawal in pendingWithdrawals.values():
            subtractfeefrom.append(pendingWithdrawal.destination_address)
            existingAmount = amounts.get(pendingWithdrawal.destination_address, 0.0)
            amounts[pendingWithdrawal.destination_address] = (pendingWithdrawal.amount / SATOSHI_PER_BITCOIN) + existingAmount
        
        OnboardingLogger(f"broadcastTransaction: {amounts}")
        try:
            txid = await self.client.sendmany(amounts, minconf=1, subtractfeefrom=subtractfeefrom)
            OnboardingLogger(f"/broadcastTransaction: {txid}")
            return txid
        except Exception as e:
            OnboardingLogger(e)
            raise

    async def getConfirmedTransactions(self, lastblockhash: str = '') -> ListSinceBlockResponse:
        targetConfirmations = self.getTargetConfirmations() if lastblockhash else 1
        OnboardingLogger(f'lastblockhash: {lastblockhash}, targetConfirmations: {targetConfirmations}')
        
        try:
            response = await self.client.listsinceblock(lastblockhash, targetConfirmations)
            return response
        except Exception as e:
            OnboardingLogger(f"listsinceblock error: {e}")
            raise

    async def getTransaction(self, transaction_id: str) -> GetTransactionResponse:
        try:
            return await self.client.gettransaction(transaction_id)
        except Exception as e:
            OnboardingLogger(e)
            raise
    
    async def getAddressGroupings(self) -> List[List[AddressGroupingItem]]:
        try:
            return await self.client.listaddressgroupings()
        except Exception as e:
            OnboardingLogger(e)
            raise

    async def getBlockHeader(self, blockhash: str) -> GetBlockHeaderResponse:
        try:
            return await self.client.getblockheader(blockhash)
        except Exception as e:
            OnboardingLogger(e)
            raise
    
    async def getBlockHeight(self) -> int:
        try:
            return await self.client.getblockcount()
        except Exception as e:
            OnboardingLogger(e)
            raise

File: backend/layer2bridge/src/full-node-interface.ts
import {
  BitcoinRPCClient,
  isWalletAlreadyLoaded,
  type AddressGroupingEntry,
  type BitcoinRpcClient,
  type BroadcastWithdrawalInput,
  type GetBlockHeaderResult,
  type GetTransactionResult,
  type ListSinceBlockResult,
  type SendManyAmounts,
  type WithdrawalTransactionOutput,
} from '@openl2/bitcoin-core-rpc';
import type { Layer2BridgeBitcoinConfFileSettings } from '@openl2/config-loader';
import { SATOSHI_PER_BITCOIN } from '@openl2/openl2-messaging';
import type { PendingWithdrawalRow } from './db/schema';

const TESTNET_TARGET_CONFIRMATIONS = 3;
const MAINNET_TARGET_CONFIRMATIONS = 6;
const DEFAULT_MINIMUM_TRANSACTION_AMOUNT = 1000;

function log(message: string): void {
  console.log(`${new Date().toISOString()} ${message}`);
}

function isTestnetChain(chain: string): boolean {
  return chain !== 'main';
}

export class BitcoinFullNodeRpc implements BitcoinRpcClient {
  private readonly client: BitcoinRPCClient;
  private readonly testnet: boolean;

  constructor(
    rpcConfig: Layer2BridgeBitcoinConfFileSettings,
    private readonly walletName: string,
  ) {
    this.client = new BitcoinRPCClient(rpcConfig, walletName);
    this.testnet = isTestnetChain(rpcConfig.chain);
    log(`bitcoin RPC wallet=${walletName} host=${rpcConfig.rpchost}:${rpcConfig.rpcport}`);
  }

  getMinimumTransactionAmount(): number {
    return DEFAULT_MINIMUM_TRANSACTION_AMOUNT;
  }

  getTargetConfirmations(): number {
    return this.testnet ? TESTNET_TARGET_CONFIRMATIONS : MAINNET_TARGET_CONFIRMATIONS;
  }

  async loadWallet(): Promise<void> {
    const response = await this.client.loadWallet(this.walletName);
    if (response.result) {
      return;
    }
    if (isWalletAlreadyLoaded(response.error)) {
      log(`wallet ${this.walletName} already loaded`);
      return;
    }
    if (response.error) {
      throw new Error(`Failed to load wallet: ${response.error.message}`);
    }
    throw new Error('Failed to load wallet: unknown error');
  }

  async getConfirmedTransactions(lastBlockHash: string): Promise<ListSinceBlockResult> {
    const targetConfirmations = lastBlockHash ? this.getTargetConfirmations() : 1;
    log(`lastblockhash: ${lastBlockHash}, targetConfirmations: ${targetConfirmations}`);
    return this.client.listSinceBlock(lastBlockHash, targetConfirmations);
  }

  async getBlockHeader(blockHash: string): Promise<GetBlockHeaderResult> {
    return this.client.getBlockHeader(blockHash);
  }

  async broadcastTransaction(pendingWithdrawals: BroadcastWithdrawalInput[]): Promise<string> {
    const amounts: SendManyAmounts = {};
    const subtractFeeFrom: string[] = [];
    for (const pendingWithdrawal of pendingWithdrawals) {
      subtractFeeFrom.push(pendingWithdrawal.destinationAddress);
      const existingAmount = amounts[pendingWithdrawal.destinationAddress] ?? 0;
      amounts[pendingWithdrawal.destinationAddress] =
        existingAmount + pendingWithdrawal.amountSatoshis / SATOSHI_PER_BITCOIN;
    }
    log(`broadcastTransaction: ${JSON.stringify(amounts)}`);
    const txid = await this.client.sendMany(amounts, {
      minconf: 1,
      subtractFeeFrom,
    });
    log(`/broadcastTransaction: ${txid}`);
    return txid;
  }

  async getTransaction(transactionId: string): Promise<GetTransactionResult> {
    return this.client.getTransaction(transactionId);
  }

  getWithdrawalOutputsFromTransaction(transaction: GetTransactionResult): WithdrawalTransactionOutput[] {
    const outputs: WithdrawalTransactionOutput[] = [];
    for (const detail of transaction.details) {
      if (!detail.address) {
        continue;
      }
      outputs.push({
        address: detail.address,
        vout: detail.vout,
        amountSatoshis: Math.round(Math.abs(detail.amount) * SATOSHI_PER_BITCOIN),
      });
    }
    return outputs;
  }

  async sendMany(outputs: SendManyAmounts): Promise<string> {
    return this.client.sendMany(outputs, { minconf: 1 });
  }

  async getAddressGroupings(): Promise<AddressGroupingEntry[]> {
    const groupings = await this.client.listAddressGroupings();
    const addresses: AddressGroupingEntry[] = [];
    for (const grouping of groupings) {
      for (const entry of grouping) {
        addresses.push({
          address: entry.address,
          amount: entry.amount,
          label: entry.label ?? '',
        });
      }
    }
    return addresses;
  }

  /** Convert bridge DB rows to RPC broadcast inputs. */
  static toBroadcastInputs(
    pendingWithdrawals: Map<string, PendingWithdrawalRow>,
  ): BroadcastWithdrawalInput[] {
    return [...pendingWithdrawals.values()].map((withdrawal) => ({
      withdrawalId: withdrawal.layer2WithdrawalId,
      destinationAddress: withdrawal.destinationAddress,
      amountSatoshis: withdrawal.amount,
    }));
  }
}
